### PR TITLE
Support using dataType param to specify edge highlighting

### PR DESCRIPTION
**Pull request by Dingzhaocheng**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?
Added property configuration to specify edge highlighting

### Before: What was the problem?
Source from this issue: https://github.com/apache/echarts/issues/16134#issuecomment-997340280, need to solve the problem of highlighting on the specified side

### After: How is it fixed in this PR?
According to the original method "highlight" of the highlighted node, pass in the dataType property in the payload

## Misc

- [x] The API has been changed (apache/echarts-doc#xxx).
- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

```js
echartInstance.dispatchAction({ type: 'highlight', dataType: 'edge | node', dataIndex: number | number[] })
```

### Merging options

- [x] Please squash the commits into a single one when merging.

### CODE REVIEW

File: src/view/Chart.ts [hunk 1]
```diff
@@ -141,18 +141,23 @@ class ChartView {
             plan: renderTaskPlan,
             reset: renderTaskReset
         });
-        this.renderTask.context = {view: this} as SeriesTaskContext;
+        this.renderTask.context = { view: this } as SeriesTaskContext;
     }
 
-    init(ecModel: GlobalModel, api: ExtensionAPI): void {}
+    init(ecModel: GlobalModel, api: ExtensionAPI): void { }
```
Comment (pissang) [inline src/view/Chart.ts:147]: Please avoid using auto formatting tool on the source code.

**Comment by pissang** [comment]
Please dont commit yarn.lock

File: src/view/Chart.ts (hunk 1 above)
Comment (Dingzhaocheng) [reply src/view/Chart.ts]: Sorry, I modified and resubmitted

**Comment by pissang** [comment]
Please don't commit dist file. Just the changed source code

File: src/view/Chart.ts [hunk 2]
```diff
@@ -152,7 +152,8 @@ class ChartView {
      * Highlight series or specified data item.
      */
     highlight(seriesModel: SeriesModel, ecModel: GlobalModel, api: ExtensionAPI, payload: Payload): void {
-        toggleHighlight(seriesModel.getData(), payload, 'emphasis');
+       const dataType = payload?.dataType || null;
```
Comment (pissang) [inline src/view/Chart.ts:155]: `|| null` seems to be not necessary

File: .gitignore [hunk 3]
```diff
@@ -17,6 +17,7 @@ local.properties
 .class
 .settings/
 .loadpath
+*.lock
```
Comment (plainheart) [inline .gitignore:20]: `yarn.lock` shouldn't be ignored even though we don't use *__yarn__*.

**Comment by pissang** [comment]
@Dingzhaocheng I didn't mean deleting these two folders. Just don't commit the change.

**Comment by Dingzhaocheng** [comment]
> @Dingzhaocheng I didn't mean deleting these two folders. Just don't commit the change.
> 
> Make sure in the `Files Changed` tab only having the changed source code and added test files.

I'm sorry, I didn't get it right earlier. 😊

File: .gitignore (hunk 3 above)
Comment (Dingzhaocheng) [reply .gitignore]: Thanks, it has been modified

File: src/view/Chart.ts (hunk 2 above)
Comment (Dingzhaocheng) [reply src/view/Chart.ts]: Thanks, it has been modified

File: src/view/Chart.ts [hunk 4]
```diff
@@ -152,7 +152,7 @@ class ChartView {
      * Highlight series or specified data item.
      */
     highlight(seriesModel: SeriesModel, ecModel: GlobalModel, api: ExtensionAPI, payload: Payload): void {
-        toggleHighlight(seriesModel.getData(), payload, 'emphasis');
+        toggleHighlight(seriesModel.getData(payload?.dataType), payload, 'emphasis');
```
Comment (pissang) [inline src/view/Chart.ts:155]: Using 

```js
payload && payload.dataType
```

Will be simpler on the generated code.

**Comment by pissang** [comment]
> I'm sorry, I didn't get it right earlier. 😊

Never mind. We are glad to having your contribution

File: src/view/Chart.ts (hunk 4 above)
Comment (Dingzhaocheng) [reply src/view/Chart.ts]: Thank you for your guidance, I have revised and submitted

File: src/view/Chart.ts [hunk 5]
```diff
@@ -152,7 +152,7 @@ class ChartView {
      * Highlight series or specified data item.
      */
     highlight(seriesModel: SeriesModel, ecModel: GlobalModel, api: ExtensionAPI, payload: Payload): void {
-        toggleHighlight(seriesModel.getData(), payload, 'emphasis');
+        toggleHighlight(seriesModel.getData(payload && payload.dataType), payload, 'emphasis');
```
Comment (pissang) [inline src/view/Chart.ts:155]: Developers may give wrong dataType that is not supported in the echarts. It's better to check if `data` exists here.

Also `downplay` method should also add this change.

File: src/view/Chart.ts [hunk 6]
```diff
@@ -152,14 +152,14 @@ class ChartView {
...
```
Comment (pissang) [inline src/view/Chart.ts:155]: If we need to do null checking. We should always do it at the place where it is used. In this case is before we passing it as a parameter to `toggleHighlight`. What if `seriesModel.getData()` has other methods to fetch the data if you do null checking in the `getLinkedData`. ? 

Here is an example:

```ts
const data = seriesModel.getData(payload && payload.dataType)
if (!data) {
    if (__DEV__) {
        error(`Unkown dataType ${dataType}`);
    }
    return;
}
toggleHighlight(data, payload, 'emphasis');
```

Null checking in the `getLinkedData` can be removed to simplify the code.

**Review by pissang — CHANGES_REQUESTED** [review]

File: src/data/helper/linkSeriesData.ts [hunk 7]
```diff
@@ -26,6 +26,7 @@ import { curry, each, assert, extend, map, keys } from 'zrender/src/core/util';
 import SeriesData from '../SeriesData';
 import { makeInner } from '../../util/model';
 import { SeriesDataType } from '../../util/types';
+import {warn} from "../../util/log";
```
Comment (pissang) [inline src/data/helper/linkSeriesData.ts:29]: Unused code here

**Review by pissang — APPROVED** [review]

**Comment by maneetgoyal** [comment]
@Dingzhaocheng hi, how can we highlight both nodes and edges simultaneously?

**Comment by maneetgoyal** [comment]
@pissang please help

**Comment by Dingzhaocheng** [comment]
> @pissang please help

You can see the PR changes file, you can follow these changes to directly modify your local package, then you can use the new API on the proposal, the test case also has the usage, you can refer to it.

**Comment by Dingzhaocheng** [comment]
> @Dingzhaocheng hi, how can we highlight both nodes and edges simultaneously?

You can see the PR changes file, you can follow these changes to directly modify your local package, then you can use the new API on the proposal, the test case also has the usage, you can refer to it

**Comment by Dingzhaocheng** [comment]
> > @Dingzhaocheng hi, how can we highlight both nodes and edges simultaneously?
> 
> You can see the PR changes file, you can follow these changes to directly modify your local package, then you can use the new API on the proposal, the test case also has the usage, you can refer to it

I'm not sure if the new version already has this method, you can check the documentation, if so, then just update to the new version

**Comment by maneetgoyal** [comment]
@Dingzhaocheng i am using the latest version. The "dataType" property is working but I want to highlight both nodes and edges at the same time.

Currently it is allowing me to highlight only one of them at a time.

Can you suggest something?

**Comment by maneetgoyal** [comment]
@pissang @Ovilia I can raise a PR if you guys or others from the Echarts team can review it?

**Pull request merged**